### PR TITLE
[air] Add batch predictor class

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

What: This class adds a generic `BatchPredictor` class that offers an interface to run batch inference on Ray datasets. It takes a Predictor class and checkpoint as an input, and provides a ``predict(dataset)`` method to run scalable scoring inference.

Why: Currently users have to implement scorers themselves. This is mostly boilerplate and prone to errors, so we should provide a simple solution instead.

Note that this predictor also implements the `Predictor` interface.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Yard1** [comment]
I don't really like the `score` nomenclature here. Scoring implies that you will receive a score - a single value. Can we do `BatchPredictor`, or something similar?

File: python/ray/ml/tests/test_predictor.py [hunk 1]
```diff
@@ -26,6 +31,18 @@ def test_serialization():
         ray.put(predictor)
 
 
+def test_scoring():
+    scorer = BatchScorer(DummyPredictor, Checkpoint.from_dict({"factor": 2.0}))
+
+    test_dataset = ray.data.from_items([1.0, 2.0, 3.0, 4.0])
+    assert scorer.score(test_dataset).to_pandas().to_numpy().squeeze().tolist() == [
```
Comment (clarkzinzow) [inline python/ray/ml/tests/test_predictor.py:38]: If looking for a more concise conversion + comparison, content should be assertable without this chain, something like:
```python
assert scorer.score(test_dataset).take() == [2.0, 4.0, 6.0, 8.0]
```

Comment (krfricke) [reply python/ray/ml/tests/test_predictor.py:38]: This leads to 

```
E       AssertionError: assert [{'value': 8.0}, {'value': 6.0}, {'value': 4.0}, {'value': 2.0}] == [2.0, 4.0, 6.0, 8.0]
```

how do I get them as a series and not dicts?

Also, the order problem seems to remain, any insights on that?

File: python/ray/ml/scorer.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        **predict_kwargs
+    ) -> ray.data.Dataset:
+        """Run batch scoring on dataset.
+
+        Args:
+            dataset: Ray dataset to run batch scoring on.
+            batch_size: Split dataset into batches of this size for prediction.
+            max_scoring_actors: If set, specify the maximum number of scoring actors.
+            ray_remote_args: Additional resource requirements to request from
+                ray (e.g., num_gpus=1 to request GPUs for the map tasks).
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/ml/scorer.py]: Agreed, this should be easily accessible. I've added it to the top level API

**Comment by krfricke** [comment]
Regarding the naming, I don't think that "scoring" implies it's only 1 measure, but yeah, it's not ideal because we're not actually scoring a dataset, but perform inference on it. 

I wouldn't want to go with BatchPredictor or anything *Predictor, as we do have a top-level Predictor interface and this wrapper does not implement it.

Maybe `BatchInference`?

**Comment by Yard1** [comment]
`BatchInference` is better, yeah.

**Comment by krfricke** [comment]
Actually, we can go with `BatchPredictor` if we implement the Predictor interface - which works quite well here tbh. I've changed this in the last commit, let me know what you think

**Comment by Yard1** [comment]
`BatchPredictor` is even better :)

File: python/ray/ml/predictor.py [hunk 3]
```diff
@@ -51,3 +52,90 @@ def __reduce__(self):
...
+            checkpoint=checkpoint, predictor_cls=predictor_cls, **kwargs
+        )
+
+    def predict(
+        self,
+        data: ray.data.Dataset,
+        *,
+        batch_size: int = 4096,
+        max_scoring_actors: Optional[int] = None,
+        num_gpus: int = 0,
```
Comment (Yard1) [inline python/ray/ml/predictor.py:97]: what about other types of resources?

Comment (krfricke) [reply python/ray/ml/predictor.py:97]: They can be specified in the resource kwargs, we've only promoted the GPUs as @clarkzinzow mentioned this is probably a very common thing to do. But I agree it's a bit weird to include GPUs and not CPUs, so I'll promote `num_cpus` as well and leave the rest in the dict.

File: python/ray/ml/tests/test_predictor.py [hunk 4]
```diff
@@ -26,6 +35,23 @@ def test_serialization():
         ray.put(predictor)
 
 
+def test_batch_prediction():
+    batch_predictor = BatchPredictor.from_checkpoint(
+        Checkpoint.from_dict({"factor": 2.0}), DummyPredictor
+    )
+
+    test_dataset = ray.data.from_items([1.0, 2.0, 3.0, 4.0])
+    # Todo: Remove sorted after https://github.com/ray-project/ray/issues/23833 fixed
```
Comment (krfricke) [inline python/ray/ml/tests/test_predictor.py:44]: There's an ordering bug tracked in https://github.com/ray-project/ray/issues/23833
Since this is a general error and not part of this PR, I'm relaxing the test constraints here.

cc @clarkzinzow can you remove this when https://github.com/ray-project/ray/issues/23833 is fixed?

File: python/ray/ml/predictor.py [hunk 5]
```diff
@@ -51,3 +52,93 @@ def __reduce__(self):
             "to serialize a checkpoint and initialize the Predictor with "
             "Predictor.from_checkpoint."
         )
+
+
+class BatchPredictor(Predictor):
```
Comment (ericl) [inline python/ray/ml/predictor.py:57]: Could you move this into a separate batch_predictor.py file?

File: python/ray/ml/predictor.py [hunk 6]
```diff
@@ -51,3 +52,93 @@ def __reduce__(self):
...
+        return BatchPredictor(
+            checkpoint=checkpoint, predictor_cls=predictor_cls, **kwargs
+        )
+
+    def predict(
+        self,
+        data: ray.data.Dataset,
+        *,
+        batch_size: int = 4096,
+        max_scoring_actors: Optional[int] = None,
```
Comment (ericl) [inline python/ray/ml/predictor.py:96]: Could we also add min_scoring_actors, so that the user can setup a fixed size pool if needed?

Comment (Yard1) [reply python/ray/ml/predictor.py:96]: at this point we may as well pass in a scaling config 🤔

File: python/ray/ml/tests/test_predictor.py (hunk 4 above)
Comment (clarkzinzow) [reply python/ray/ml/tests/test_predictor.py:44]: @krfricke fixing [PR](https://github.com/ray-project/ray/pull/23837) is out FYI!

**Review by ericl — APPROVED** [review]

File: doc/source/ray-air/getting-started.rst [hunk 7]
```diff
@@ -89,6 +89,9 @@ Predictors
 .. autoclass:: ray.ml.predictor.Predictor
     :members:
 
+.. autoclass:: ray.ml.batch_predictor.BatchPredictor
```
Comment (ericl) [inline doc/source/ray-air/getting-started.rst:92]: :+1:
